src/PointwiseFunctions/GeneralRelativity/Psi4.cpp

```diff
@@ -0,0 +1,143 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "PointwiseFunctions/GeneralRelativity/Psi4.hpp"
+
+#include <cstddef>
+
+#include "DataStructures/LeviCivitaIterator.hpp"
+#include "DataStructures/Tensor/EagerMath/Magnitude.hpp"
+#include "DataStructures/Tensor/Expressions/Evaluate.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Domain/CoordinateMaps/Tags.hpp"
+#include "Domain/Structure/Element.hpp"
+#include "Domain/Tags.hpp"
+#include "PointwiseFunctions/GeneralRelativity/ProjectionOperators.hpp"
+#include "PointwiseFunctions/GeneralRelativity/WeylPropagating.hpp"
+#include "Utilities/ContainerHelpers.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/MakeWithValue.hpp"
+
+namespace gr {
+template <typename ComplexDataType, typename Frame, typename RealDataType>
+void psi_4(
+    const gsl::not_null<Scalar<ComplexDataType>*> psi_4_result,
+    const tnsr::ii<RealDataType, 3, Frame>& spatial_ricci,
+    const tnsr::ii<RealDataType, 3, Frame>& extrinsic_curvature,
+    const tnsr::ijj<RealDataType, 3, Frame>&
+        cov_deriv_extrinsic_curvature,
+    const tnsr::ii<RealDataType, 3, Frame>& spatial_metric,
+    const tnsr::II<RealDataType, 3, Frame>& inverse_spatial_metric,
+    const tnsr::I<RealDataType, 3, Frame>& inertial_coords) {
+  const auto magnitude_cartesian =
+      Scalar<RealDataType>(magnitude(inertial_coords, spatial_metric));
+  auto r_hat = make_with_value<tnsr::I<RealDataType, 3, Frame>>(
+      get<0, 0>(spatial_metric), 0.0);
+  for (size_t j = 0; j < 3; j++) {
+    if constexpr (is_derived_of_vector_impl_v<RealDataType>) {
+      for (size_t i = 0; i < magnitude_cartesian.size(); i++) {
+        if (magnitude_cartesian.get()[i] != 0.0) {
+          r_hat.get(j)[i] =
+              inertial_coords.get(j)[i] / magnitude_cartesian.get()[i];
+        }
+      }
+    } else {
+      if (magnitude_cartesian.get() != 0.0) {
+        r_hat.get(j) = inertial_coords.get(j) / magnitude_cartesian.get();
+      }
+    }
+  }
+  const auto lower_r_hat =
+      tenex::evaluate<ti::i>(r_hat(ti::J) * spatial_metric(ti::i, ti::j));
+
+  const auto projection_tensor =
+      transverse_projection_operator(spatial_metric, lower_r_hat);
+  const auto inverse_projection_tensor =
+      transverse_projection_operator(inverse_spatial_metric, r_hat);
+  const auto projection_up_lo = tenex::evaluate<ti::K, ti::i>(
+      projection_tensor(ti::i, ti::j) * inverse_spatial_metric(ti::K, ti::J));
+
+  const auto u8_plus = gr::weyl_propagating(
+      spatial_ricci, extrinsic_curvature, inverse_spatial_metric,
+      cov_deriv_extrinsic_curvature, r_hat, inverse_projection_tensor,
+      projection_tensor, projection_up_lo, 1.0);
+
+  auto x_coord = make_with_value<tnsr::I<RealDataType, 3, Frame>>(
+      get<0, 0>(inverse_spatial_metric), 0.0);
+  x_coord.get(0) = 1.0;
+  const auto magnitude_x =
```

Comment:
no need to allocate magnitude_x or x_coord here...just put these right into line 71, and similarly for y_hat